src/coreclr/jit/optimizer.cpp

```diff
@@ -8544,15 +8558,20 @@ bool Compiler::optExtractArrIndex(GenTree* tree, ArrIndex* result, unsigned lhsN
     {
         return false;
     }
-    if (si->gtOper != GT_LSH)
+    GenTree* index;
+    if (si->gtOper == GT_LSH)
     {
-        return false;
+        GenTree* scale = si->gtGetOp2();
+        index          = si->gtGetOp1();
+        if (scale->gtOper != GT_CNS_INT)
+        {
+            return false;
+        }
     }
-    GenTree* scale = si->gtGetOp2();
-    GenTree* index = si->gtGetOp1();
-    if (scale->gtOper != GT_CNS_INT)
+    else
     {
-        return false;
+        // No scale (e.g., byte array).
+        index = si;
```

Comment:
Do we need to add some assert to make sure the array is what we expect?

Reply:
Sorry, I missed this before merge. There's certainly some question about what additional things could/should be checked in this pattern matching. Types, for example. For this case, the index is checked (lower down) to be a lclvar that is the same as the array bounds check base, so we're good.